host.go

```diff
@@ -53,81 +97,89 @@ func newRunner(m manifest.ManifestProvider, client *client) (*gpbftRunner, error
 	return runner, nil
 }
 
-func (h *gpbftRunner) Run(instance uint64, ctx context.Context) (_err error) {
-	h.runningCtx, h.ctxCancel = context.WithCancel(ctx)
-	defer h.ctxCancel()
+func (h *gpbftRunner) Start(ctx context.Context) (_err error) {
+	defer func() {
+		if _err != nil {
+			_err = multierr.Append(_err, h.Stop(ctx))
+		}
+	}()
 
-	if err := h.participant.StartInstanceAt(instance, time.Now()); err != nil {
+	startInstance := h.manifest.InitialInstance
+	if latest := h.certStore.Latest(); latest != nil {
+		startInstance = latest.GPBFTInstance + 1
+	}
+
+	messageQueue, err := h.startPubsub()
+	if err != nil {
+		return err
+	}
+
+	if err := h.participant.StartInstanceAt(startInstance, time.Now()); err != nil {
 		return xerrors.Errorf("starting a participant: %w", err)
 	}
 
 	// Subscribe to new certificates. We don't bother canceling the subscription as that'll
 	// happen automatically when the channel fills.
 	finalityCertificates := make(chan *certs.FinalityCertificate, 4)
-	_, _ = h.client.certStore.SubscribeForNewCerts(finalityCertificates)
-
-	defer func() {
-		if _err != nil {
-			h.log.Errorf("gpbfthost exiting: %+v", _err)
-		}
-	}()
-
-	messageQueue := h.client.IncomingMessages()
-	for {
+	_, _ = h.certStore.SubscribeForNewCerts(finalityCertificates)
 
-		// prioritise finality certificates and alarm delivery
-		select {
-		case c, ok := <-finalityCertificates:
-			if !ok {
-				finalityCertificates = make(chan *certs.FinalityCertificate, 4)
-				c, _ = h.client.certStore.SubscribeForNewCerts(finalityCertificates)
-			}
-			if err := h.receiveCertificate(c); err != nil {
-				return err
+	h.errgrp.Go(func() (_err error) {
+		defer func() {
+			// TODO: catch panics
+			if _err != nil && h.runningCtx.Err() == nil {
+				h.log.Errorf("exited GPBFT runner early: %+v", _err)
 			}
-			continue
-		case <-h.alertTimer.C:
-			if err := h.participant.ReceiveAlarm(); err != nil {
-				return err
+		}()
+		for h.runningCtx.Err() == nil {
+			// prioritise finality certificates and alarm delivery
+			select {
+			case c, ok := <-finalityCertificates:
+				if !ok {
+					finalityCertificates = make(chan *certs.FinalityCertificate, 4)
+					c, _ = h.certStore.SubscribeForNewCerts(finalityCertificates)
+				}
+				if err := h.receiveCertificate(c); err != nil {
+					return err
+				}
+				continue
+			case <-h.alertTimer.C:
+				if err := h.participant.ReceiveAlarm(); err != nil {
+					return err
```

Comment:
ignore this error?

Reply:
I don't think we can, I think error here would suggest that gpbft is in bad state.